iocore/eventsystem/UnixEventProcessor.cc

```diff
@@ -134,10 +153,77 @@ ThreadAffinityInitializer::set_affinity(int, Event *)
   } else {
     Warning("hwloc returned an unexpected number of objects -- CPU affinity disabled");
   }
-#endif // TS_USE_HWLOC
   return 0;
 }
 
+void *
+ThreadAffinityInitializer::alloc_numa_stack(EThread *t, size_t stacksize)
+{
+  hwloc_membind_policy_t mem_policy = HWLOC_MEMBIND_DEFAULT;
+  hwloc_nodeset_t nodeset           = hwloc_bitmap_alloc();
+  int num_nodes                     = 0;
+  void *stack                       = nullptr;
+  hwloc_obj_t obj                   = hwloc_get_obj_by_type(ink_get_topology(), obj_type, t->id % obj_count);
+
+  // Find the NUMA node set that correlates to our next thread CPU set
+  hwloc_cpuset_to_nodeset(ink_get_topology(), obj->cpuset, nodeset);
+  // How many NUMA nodes will we be needing to allocate across?
+  num_nodes = hwloc_get_nbobjs_inside_cpuset_by_type(ink_get_topology(), obj->cpuset, HWLOC_OBJ_NODE);
+
+  if (num_nodes == 1) {
+    // The preferred memory policy. The thread lives in one NUMA node.
+    mem_policy = HWLOC_MEMBIND_BIND;
+  } else if (num_nodes > 1) {
+    // If we have mode than one NUMA node we should interleave over them.
+    mem_policy = HWLOC_MEMBIND_INTERLEAVE;
+  }
+
+  if (mem_policy != HWLOC_MEMBIND_DEFAULT) {
+    // Let's temporarily set the memory binding to our destination NUMA node
+    hwloc_set_membind_nodeset(ink_get_topology(), nodeset, mem_policy, HWLOC_MEMBIND_THREAD);
+  }
+
+  // Alloc our stack
+  stack = this->alloc_hugepage_stack(stacksize);
+
+  if (mem_policy != HWLOC_MEMBIND_DEFAULT) {
+    // Now let's set it back to default for this thread.
+    hwloc_set_membind_nodeset(ink_get_topology(), hwloc_topology_get_topology_nodeset(ink_get_topology()), HWLOC_MEMBIND_DEFAULT,
+                              HWLOC_MEMBIND_THREAD);
+  }
+
+  hwloc_bitmap_free(nodeset);
+
+  return stack;
+}
+
+void *
+ThreadAffinityInitializer::alloc_stack(EThread *t, size_t stacksize)
+{
+  return this->obj_count > 0 ? this->alloc_numa_stack(t, stacksize) : this->alloc_hugepage_stack(stacksize);
```

Comment:
I had to read this numerous times, it was not obvious. In the end, this is exactly the same as doing:

```
return  this->alloc_numa_stack(t, stacksize)
```

right? I understand (since you told me) that it might be a slight optimization, but is that worth it for something that happens once at startup, at the cost of more confusing code ?